### PR TITLE
Read version from Cargo.toml

### PR DESCRIPTION
**Pull request by ruphin**
Read the version from Cargo.toml using an `env!()` macro

### CODE REVIEW

**Comment by F001** [comment]
At present, this can't work. There is a build failure:

> error: environment variable `CARGO_PKG_VERSION` not defined

I guess it depends on #695

**Comment by ruphin** [comment]
It happens because we are not using Cargo to build, so the env variable isn't properly set. I'm working on a patch in `build_extra/rust/run.py` that sets it.

**Comment by ruphin** [comment]
Ok, looking into it.

**Comment by ruphin** [comment]
I've done some searching, but I can't figure out how to set environment variables in GN. I could add it as an argument, and parse the args and set the env variable in `run.py`, but that does not feel like a clean solution. I can implement this but I'm not sure it's worth it.

The main goal of this patch was to get compatibility with Cargo. If we're switching to Cargo in the future I don't think there is much value in adding micro-optimisations to the build at this point.

**Comment by ry** [comment]
@ruphin You can't set env vars in GN. Pass the version as an argument to https://github.com/denoland/deno/blob/master/build_extra/rust/run.py you can set the env there.

**Comment by ry** [comment]
Any progress on this one? It would be nice to have and it's almost complete.

**Comment by ruphin** [comment]
I've been busy renovating my house, sorry for the delay. I'll get it in
before end of this week.

**Comment by ruphin** [comment]
I think this should work, but the CI breaks on format rules. It doesn't output what's wrong so I don't know what to change to fix it.

**Comment by ry** [comment]
tools/format.py should fix that

File: build_extra/rust/get_cargo_info.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env python
+# Copyright 2018 the Deno authors. All rights reserved. MIT license.
+
+import os
+import re
+
+# Read the package version from Cargo.toml and output as json
+current_path = os.path.dirname(os.path.realpath(__file__))
+for line in open(os.path.normpath(current_path + "/../../Cargo.toml")):
```
Comment (F001) [inline build_extra/rust/get_cargo_info.py]: `os.path.join(current_path, "../../Cargo.toml")` is better than string concatenation.

Comment (ruphin) [reply build_extra/rust/get_cargo_info.py]: Thanks for the suggestion, I'll update.

**Comment by ruphin** [comment]
It's all fixed now

**Comment by ry** [comment]
To make sure deno.exe rebuilds when Cargo.toml is updated, I've modified the patch so that `get_cargo_info.py` is called in `//BUILD.gn` instead of in `rust.gni`.

**Pull request merged**